### PR TITLE
Error(Table) to support coercion

### PR DESCRIPTION
**Pull request by anderson-joyle**
Issue https://github.com/microsoft/Power-Fx/issues/1250.

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Public/ErrorKind.cs [hunk 1]
```diff
@@ -13,8 +13,8 @@ public enum ErrorKind
         Sync = 1,
         MissingRequired = 2,
         CreatePermission = 3,
-        EditPermission = 4,
-        DeletePermission = 5,
+        EditPermissions = 4,
+        DeletePermissions = 5,
```
Comment (anderson-joyle) [inline src/libraries/Microsoft.PowerFx.Core/Public/ErrorKind.cs:17]: These enums were diverging from PA Client.
[https://github.com/microsoft/Power-Fx/blob/a75cc231f608e928e9cda5f44aee167625d82881/src/libraries/Microsoft.PowerFx.Core/Types/Enums/EnumStoreBuilder.cs#L46](https://github.com/microsoft/Power-Fx/blob/a75cc231f608e928e9cda5f44aee167625d82881/src/libraries/Microsoft.PowerFx.Core/Types/Enums/EnumStoreBuilder.cs#L46)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs [hunk 2]
```diff
@@ -2251,6 +2251,17 @@ public static FormulaValue Error(IRContext irContext, FormulaValue[] args)
...
+                        if (!int.TryParse(sv.Value, out var kindInt))
+                        {
+                            return CommonErrors.RuntimeTypeMismatch(irContext);
+                        }
+
+                        errorKind = (ErrorKind)kindInt;
+                        break;
+                    case BooleanValue bv:
+                        errorKind = (ErrorKind)Convert.ToInt32(bv.Value, CultureInfo.InvariantCulture);
+                        break;
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:2264]: I know this is the way it's currently implemented, but can this be solved in the binder, by using coercion nodes? For example, something like `Error( [{Kind:ErrorKind.Div0}, {Kind:"14"}] )` be IR'd as

    Error:N(
        Table:*[Kind:n](
            {Kind: OptionSetToNumber:n(ScopeAccess(Scope1,Kind))},
            {Kind: Value:n("14")}))

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Comment by anderson-joyle** [comment]
>             }

Unused code?

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Error.cs:99 in 060a73c. [](commit_id = 060a73c7f6244dab1df3576ae74e4d984bb259a3, deletion_comment = False)

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Error.txt [hunk 3]
```diff
@@ -123,4 +123,31 @@ Error(Table({Kind:ErrorKind.Div0},{Kind:ErrorKind.Numeric}))
...
+Error({Kind:ErrorKind.Sync})
+
+>> Error ({Kind:"4"})
+Error({Kind:ErrorKind.EditPermissions})
+
+>> IfError(Error (Table({Kind:0,Message:"Custom error message NONE"}, {Kind:1,Message:"Custom error message SYNC"})), Concat(AllErrors,$"{Kind}:{Message}",", "))
+"0:Custom error message NONE, 1:Custom error message SYNC"
+
+>> IfError(Error(Table({Kind:0,Message:"Custom error message NONE"}, {Kind:1,Message:"Custom error message SYNC"})),Concat(AllErrors,Message,","))
+"Custom error message NONE,Custom error message SYNC"
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Error.txt:153]: Suggestion for tests:

    // Coercion from string, with message
    >> IfError(Error([{Kind:"12"},{Kind:"13",Message:"With message"}]), Concat(AllErrors, $"{Kind}:{Message}", ", "))
    "12:, 13:With message"

    // Coercion from string, with coerced message
    >> IfError(Error([{Kind:"12",Message:333},{Kind:"13",Message:444}]), Concat(AllErrors, $"{Kind}:{Message}", ", "))
    "12333:, 13:444"

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Review by MikeStall — DISMISSED** [review]
:shipit:

**Review by MikeStall — APPROVED** [review]

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.DeletePermission<br />❌CP0002 F:Microsoft.PowerFx.ErrorKind.EditPermission

**Pull request merged**